File: functions/src/geo.js
// prototype/src/geo.js と同一（技術検証スプリントで速度・妥当性を確認済みのロジックをそのまま本番へ移植）。
// 座標計算の共通ユーティリティ（Haversine距離、度→ラジアン変換等）
const EARTH_RADIUS_M = 6_371_000;

export function toRad(deg) {
  return (deg * Math.PI) / 180;
}

export function toDeg(rad) {
  return (rad * 180) / Math.PI;
}

/** 2点間の距離（メートル） */
export function haversineDistanceM([lat1, lon1], [lat2, lon2]) {
  const dLat = toRad(lat2 - lat1);
  const dLon = toRad(lon2 - lon1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLon / 2) ** 2;
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return EARTH_RADIUS_M * c;
}

/** 2点の中点（単純平均。区間程度の短距離であれば十分な近似） */
export function midpoint([lat1, lon1], [lat2, lon2]) {
  return [(lat1 + lat2) / 2, (lon1 + lon2) / 2];
}

/** 点から線分（2点間）への最短距離（メートル、簡易平面近似） */
export function pointToSegmentDistanceM([plat, plon], [alat, alon], [blat, blon]) {
  // 緯度経度をローカルなメートル平面に投影して計算（短距離前提の近似）
  const latRef = toRad(alat);
  const mPerDegLat = 111_320;
  const mPerDegLon = 111_320 * Math.cos(latRef);

  const px = (plon - alon) * mPerDegLon;
  const py = (plat - alat) * mPerDegLat;
  const bx = (blon - alon) * mPerDegLon;
  const by = (blat - alat) * mPerDegLat;

  const segLenSq = bx * bx + by * by;
  let t = segLenSq === 0 ? 0 : (px * bx + py * by) / segLenSq;
  t = Math.max(0, Math.min(1, t));
  const projX = bx * t;
  const projY = by * t;
  return Math.hypot(px - projX, py - projY);
}

File: app/lib/services/road_graph_engine/geo.dart
import 'dart:math' as math;

/// prototype/src/geo.js のDart移植（技術検証プロトタイプで速度・妥当性を確認済みのロジック）。
const double _earthRadiusM = 6371000;

double toRad(double deg) => deg * math.pi / 180;

double toDeg(double rad) => rad * 180 / math.pi;

double haversineDistanceM(
  double lat1,
  double lon1,
  double lat2,
  double lon2,
) {
  final dLat = toRad(lat2 - lat1);
  final dLon = toRad(lon2 - lon1);
  final a = math.pow(math.sin(dLat / 2), 2) +
      math.cos(toRad(lat1)) * math.cos(toRad(lat2)) * math.pow(math.sin(dLon / 2), 2);
  final c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a));
  return _earthRadiusM * c;
}

/// prototype/src/geo.js の pointToSegmentDistanceM のDart移植。
/// 点(plat,plon)から線分(alat,alon)-(blat,blon)への最短距離（メートル、簡易平面近似）。
/// 投稿UIの軌跡→道路区間スナップ（snapTraceToRoad）で使用。
double pointToSegmentDistanceM(
  double plat,
  double plon,
  double alat,
  double alon,
  double blat,
  double blon,
) {
  final latRef = toRad(alat);
  const mPerDegLat = 111320.0;
  final mPerDegLon = 111320.0 * math.cos(latRef);

  final px = (plon - alon) * mPerDegLon;
  final py = (plat - alat) * mPerDegLat;
  final bx = (blon - alon) * mPerDegLon;
  final by = (blat - alat) * mPerDegLat;

  final segLenSq = bx * bx + by * by;
  var t = segLenSq == 0 ? 0.0 : (px * bx + py * by) / segLenSq;
  t = t.clamp(0.0, 1.0);
  final projX = bx * t;
  final projY = by * t;
  return math.sqrt(math.pow(px - projX, 2) + math.pow(py - projY, 2));
}
